### PR TITLE
New config API, `wrangler.json` output and updated preview mode

### LINKED ISSUE #72: Add build tests for node-compat/worker-basic playground
These tests pass in dev but not when the build output is tested. Needs investigating.

### LINKED ISSUE #77: Output `wrangler.json` as part of build process

### PR DESCRIPTION
**Pull request by jamesopstad**
resolves #77
resolves #72

This PR makes the following changes:
- The config now uses a new API that reads from the user's Wrangler config file(s)
- `wrangler.json` files are now generated and included as part of the build
- Preview mode has been updated to read from the generated `wrangler.json` file(s)

Notes:
- This currently uses a prerelease version of Wrangler that is created from this PR - https://github.com/cloudflare/workers-sdk/pull/7256
- A user Wrangler config file is currently required for all projects and it must include either an `assets` field or `name` and `main` fields. We may wish to add support for starting without a Wrangler config file for assets only projects.
- We currently enforce that the output for all packages is inside the same build directory. We may wish to allow more flexibility here but this requires being able to read environment specific build options in preview mode.
- Additional validation of the input Wrangler config (e.g. to warn on redundant fields) should be added in a follow up PR.
- Once this PR is merged, I'll do a follow up to remove the `invariant` function and use `assert` instead.

### CODE REVIEW

File: packages/vite-plugin-cloudflare/src/index.ts
Comment (petebacondarwin) [inline packages/vite-plugin-cloudflare/src/index.ts:161]: I assume that this is configurable by the user?
Or maybe by other plugins?

So I would expect we should be able to grab this from the config somewhere...

File: packages/vite-plugin-cloudflare/src/index.ts
Comment (petebacondarwin) [inline packages/vite-plugin-cloudflare/src/index.ts:193]: Out of curiosity - why is this no longer needed?

File: packages/vite-plugin-cloudflare/src/miniflare-options.ts
Comment (petebacondarwin) [inline packages/vite-plugin-cloudflare/src/miniflare-options.ts:30]: Do we ever call this with `persistState` set to false?

File: packages/vite-plugin-cloudflare/src/index.ts
Comment (jamesopstad) [reply packages/vite-plugin-cloudflare/src/index.ts:161]: Yeah, I had a quick look but it wasn't obvious to me where to find this. Will look into it more.

File: packages/vite-plugin-cloudflare/src/index.ts
Comment (jamesopstad) [reply packages/vite-plugin-cloudflare/src/index.ts:193]: I think this is only relevant when you have other plugins that add middleware so I thought it's best to remove it for now and revisit when we look at framework integrations.

File: playground/node-compat/worker-crypto/wrangler.toml
Comment (dario-piotrowicz) [inline playground/node-compat/worker-crypto/wrangler.toml:1]:
```suggestion
name = "worker-crypto"
```

File: packages/vite-plugin-cloudflare/src/miniflare-options.ts
Comment (jamesopstad) [reply packages/vite-plugin-cloudflare/src/miniflare-options.ts:30]: Not at the moment. I would be inclined to add `persistState` as an option to the plugin config so left the option open. I think people didn't feel this was necessary when we discussed it the other day. Happy to go with what you prefer here.

No need for this function to be exported so I'll remove that.

File: playground/static-mpa/vite.config.ts
Comment (dario-piotrowicz) [inline playground/static-mpa/vite.config.ts:15]: why this change?

File: packages/vite-plugin-cloudflare/src/index.ts
Comment (dario-piotrowicz) [inline packages/vite-plugin-cloudflare/src/index.ts:104]: shall we create a gh issue?

File: packages/vite-plugin-cloudflare/src/plugin-config.ts
Comment (dario-piotrowicz) [inline packages/vite-plugin-cloudflare/src/plugin-config.ts:65]: why is this? 😕

File: packages/vite-plugin-cloudflare/src/plugin-config.ts
Comment (dario-piotrowicz) [inline packages/vite-plugin-cloudflare/src/plugin-config.ts:114]: do we know why environment names don't allow `-`? maybe that's something that we can change upstream? (it is pretty minor in any case 😅)

Comment (dario-piotrowicz) [inline packages/vite-plugin-cloudflare/src/plugin-config.ts:114]:
```suggestion
// Worker names can only contain alphanumeric characters and '-' whereas environment names can only contain alphanumeric characters and '$', '_'
```

File: packages/vite-plugin-cloudflare/src/plugin-config.ts
Comment (dario-piotrowicz) [inline packages/vite-plugin-cloudflare/src/plugin-config.ts:116]: nit
```suggestion
	return workerName.replaceAll('-', '_');
```

File: packages/vite-plugin-cloudflare/src/plugin-config.ts
Comment (dario-piotrowicz) [inline packages/vite-plugin-cloudflare/src/plugin-config.ts:132]:
```suggestion
		`Config not found. Have you created a wrangler.json(c) or wrangler.toml file?`,
```

**Review by dario-piotrowicz — COMMENTED** [review]
Looks good to me 😄 

I just left a minor few comments 🙂 

By the way, I am not a big fan on relying on a wrangler prerelease... that feels super brittle to me.... but this PR can be quite a blocker for other work, so I think it'll be ok this time 😕  (but I think we should not merge this PR until https://github.com/cloudflare/workers-sdk/pull/7256 is at least approved)

**Comment by jamesopstad** [comment]
> By the way, I am not a big fan on relying on a wrangler prerelease... that feels super brittle to me.... but this PR can be quite a blocker for other work, so I think it'll be ok this time 😕 (but I think we should not merge this PR until [cloudflare/workers-sdk#7256](https://github.com/cloudflare/workers-sdk/pull/7256) is at least approved)

I wouldn't say I'm a fan either but the two are intrinsically linked. The approach in this PR needs to be approved before the other PR can be moved forward and merged. I agree it would be best to merge the other PR before this one.

File: playground/module-resolution/vite.config.no-prebundling.ts
Comment (jamesopstad) [reply playground/module-resolution/vite.config.no-prebundling.ts:30]: Not sure what you mean by 'inline the env configs'. Can you explain?

I do think it might be good for us to allow setting some Vite environment options in the plugin config but that seems like something different.

File: playground/node-compat/worker-basic/wrangler.toml
Comment (jamesopstad) [reply playground/node-compat/worker-basic/wrangler.toml:1]: What's the reason for changing all these worker names? Does it matter?

File: playground/static-mpa/vite.config.ts
Comment (jamesopstad) [reply playground/static-mpa/vite.config.ts:15]: Completely inconsequential. I just did it because I didn't like how it looked 😂.

File: packages/vite-plugin-cloudflare/src/index.ts
Comment (jamesopstad) [reply packages/vite-plugin-cloudflare/src/index.ts:104]: #85

File: packages/vite-plugin-cloudflare/src/plugin-config.ts
Comment (jamesopstad) [reply packages/vite-plugin-cloudflare/src/plugin-config.ts:65]: It's the way `readConfig` from Wrangler works, unfortunately. I was hoping that if an environment is specified and no environments are present in the config file then it would just use the top level config. This doesn't seem to be the case though.

File: packages/vite-plugin-cloudflare/src/plugin-config.ts
Comment (jamesopstad) [reply packages/vite-plugin-cloudflare/src/plugin-config.ts:114]: I assume there's probably a reason. Not sure why though.

Comment (jamesopstad) [reply packages/vite-plugin-cloudflare/src/plugin-config.ts]: 89e3ff4

File: packages/vite-plugin-cloudflare/src/plugin-config.ts
Comment (jamesopstad) [reply packages/vite-plugin-cloudflare/src/plugin-config.ts]: 89e3ff4

File: packages/vite-plugin-cloudflare/src/plugin-config.ts
Comment (jamesopstad) [reply packages/vite-plugin-cloudflare/src/plugin-config.ts]: 89e3ff4

File: packages/vite-plugin-cloudflare/src/index.ts
Comment (jamesopstad) [reply packages/vite-plugin-cloudflare/src/index.ts]: 825e6d5

File: playground/module-resolution/vite.config.no-prebundling.ts
Comment (dario-piotrowicz) [reply playground/module-resolution/vite.config.no-prebundling.ts:30]: with inline env config I mean setting the environment configs in the config option here, so instead of having:
```
environments: {
		worker: {
			<CONFIGS>
		},
},
plugins: [cloudflare({ viteEnvironmentName: 'worker' })], // <- the name is used to get the <CONFIGS>
```

it would be probably feel much more natural to me to have:
```
plugins: [cloudflare({ <CONFIGS> })], // (with an optional `viteEnvironmentName`)
```
or, to have a more clear distinction between configs have a specific field for this (probably better):
```
plugins: [cloudflare({ viteEnvironmentName: 'worker', environmentConfig: <CONFIGS> })],
```

File: playground/node-compat/worker-basic/wrangler.toml
Comment (dario-piotrowicz) [reply playground/node-compat/worker-basic/wrangler.toml:1]: mh... probably not... I mean... it can make things clearer if we were to ever deploy any of these... 🤷

it is a pretty minimal thing indeed... but also I don't see why not having more specific names 🤷

not a big deal, totally up to you, if you want to keep it to `worker` that's ok by me 🙂

Comment (jamesopstad) [reply playground/node-compat/worker-basic/wrangler.toml:1]: It's fine, I'm happy to do it. I just asked because the worker names haven't changed in this PR.

**Review by dario-piotrowicz — APPROVED** [review]
Looks good to me, approving but again I think that ideally we shouldn't merge this until https://github.com/cloudflare/workers-sdk/pull/7256 is merged (or close to)

File: packages/vite-plugin-cloudflare/src/plugin-config.ts
Comment (petebacondarwin) [reply packages/vite-plugin-cloudflare/src/plugin-config.ts:65]: > I was hoping that if an environment is specified and no environments are present in the config file then it would just use the top level config

I am pretty sure that is how it works. It just warns but doesn't error: https://github.com/cloudflare/workers-sdk/blob/b5c72af7ea061cae9948ae51e3252d8857643890/packages/wrangler/src/config/validation.ts#L245

File: playground/node-compat/worker-basic/wrangler.toml
Comment (dario-piotrowicz) [reply playground/node-compat/worker-basic/wrangler.toml:1]: ah.... sorry I missed that 😅

well again, whatever you prefer I'll be fine with 🙂👍

File: packages/vite-plugin-cloudflare/src/plugin-config.ts
Comment (dario-piotrowicz) [reply packages/vite-plugin-cloudflare/src/plugin-config.ts]: https://github.com/vitejs/vite/blob/81c31c16d10bff48c9965f7aedbf7efb7c7022c4/packages/vite/src/node/baseEnvironment.ts#L57-L58

feels like a pretty arbitrary reasoning... 🤷 

but hey, we can live with that 😅 🤷

File: packages/vite-plugin-cloudflare/src/plugin-config.ts
Comment (jamesopstad) [reply packages/vite-plugin-cloudflare/src/plugin-config.ts:65]: If that was the case then you'd be able to the change `env` here from `userConfig.mode` to some arbitrary value and all the tests would still pass.

File: packages/vite-plugin-cloudflare/src/plugin-config.ts
Comment (jamesopstad) [reply packages/vite-plugin-cloudflare/src/plugin-config.ts]: Hmm, I wouldn't have thought that `-` needs escaping in directory names. It would be nicer if they could be used so maybe it is worth looking into changing.

File: packages/vite-plugin-cloudflare/src/plugin-config.ts
Comment (jamesopstad) [reply packages/vite-plugin-cloudflare/src/plugin-config.ts:64]: This comment explains the behaviour here:

https://github.com/cloudflare/workers-sdk/blob/bea65583f72fcf8346ed82d4b9b1176669bcfd23/packages/wrangler/src/config/validation.ts#L193-L205

For Workers, if an environment name is specified and the environment is not configured then a fake empty environment is used in its place. This has the effect of removing inheritable properties that are set at the top level. You can see the relevant code here (the third argument is the dummy env):

https://github.com/cloudflare/workers-sdk/blob/bea65583f72fcf8346ed82d4b9b1176669bcfd23/packages/wrangler/src/config/validation.ts#L219-L228

What we really want is the behaviour that is used for Pages but it feels like this might be a breaking change.

**Comment by jamesopstad** [comment]
It turns out that this also resolves #72. The node-compat/worker-basic build test now passes. Not entirely sure why but I'll take it 😀.

**Pull request merged**
